src/basic/log.h

```diff
@@ -367,3 +368,41 @@ int log_syntax_invalid_utf8_internal(
 #define DEBUG_LOGGING _unlikely_(log_get_max_level() >= LOG_DEBUG)
 
 void log_setup(void);
+
+typedef struct ErrorRateLimit {
```

Comment:
I think LogRateLimit would be a better name, given this is not just about error messages per se, but any kind of log message really.